Android ONNX INT8 (NNAPI/ORT)

WangHHY19931001 opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Add a complete Android C++ example for running YOLO11 INT8 ONNX on-device via NNAPI/XNNPACK with ONNX Runtime and OpenCV, including export utilities and JNI bindings.

### 📊 Key Changes
- New example: `examples/YOLO11-CPP-ONNXRUNNER-INT8/` Android Studio project with C++ JNI (Ort + OpenCV) and Java UI wrapper
- Native pipeline: image preprocessing, ONNX Runtime session init with optional NNAPI/XNNPACK, postprocessing with NMS, and JNI marshaling
- Export utilities: `export_tflite.py` (YOLO -> INT8 TFLite) and `change_onnx_mate.py` (inject ONNX metadata like names/imgsz)
- Build system: Gradle project, CMake for native code, and instructions to manually add ONNX Runtime and OpenCV Android SDKs
- README: step-by-step guide for exporting, converting TFLite->ONNX, metadata editing, and running the demo

### 🎯 Purpose & Impact
- Demonstrates a practical path to deploy YOLO11 INT8 models on Android using NNAPI/ORT for efficient on-device inference
- Provides a ready-to-build example accelerating mobile edge adoption and benchmarking
- Offers reusable JNI/C++ scaffolding and preprocessing/postprocessing patterns for downstream apps
- Highlights metadata handling for class names and imgsz to improve UX and label decoding

WangHHY19931001 committed:
Android ONNX INT8 (NNAPI/ORT)

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 commented:
I have read the CLA Document and I sign the CLA

UltralyticsAssistant commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/app/src/main/cpp/onnx_ott_runner.cpp:
```diff
@@ -0,0 +1,213 @@
+        jboolean isCopy;
+        const char* c_value = env->GetStringUTFChars(value, &isCopy);
+        std::string result(c_value);
+        if (isCopy == JNI_TRUE)
```
⚠️ **HIGH**: `ToCppString` must always call `ReleaseStringUTFChars`; skipping the release when `isCopy` is `JNI_FALSE` keeps the UTF chars pinned on the VM, leaking memory and eventually exhausting the local reference table. Please release the string unconditionally before returning.

WangHHY19931001 replied:
fix ed it

UltralyticsAssistant commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/app/src/main/java/com/example/demo_for_yolo_int8/MainActivity.java:
```diff
@@ -0,0 +1,84 @@
+        AssetManager assetManager = getAssets();
+        InputStream inputStream = null;
+        try {
+            inputStream = assetManager.open("test_2400_1080.jpg");
```
💡 **MEDIUM**: The `InputStream` from `assetManager.open(...)` is never closed, so the asset descriptor remains pinned for the lifetime of the activity. Please ensure the stream is closed (for example with try-with-resources or a finally block) once the bitmap has been decoded to avoid leaking file descriptors.

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please fix the JNI string release in `ToCppString` to prevent leaking pinned UTF chars, and close the asset `InputStream` in `MainActivity` so descriptors are not leaked during repeated runs.

💬 Posted 2 inline comments

WangHHY19931001 committed:
1. fix ToCppString must always call ReleaseStringUTFChars; skipping the release when isCopy is JNI_FALSE keeps the UTF chars pinned on the VM, leaking memory and eventually exhausting the local reference table. Please release the string unconditionally before returning.
2. fix The InputStream from assetManager.open(...) is never closed, so the asset descriptor remains pinned for the lifetime of the activity. Please ensure the stream is closed (for example with try-with-resources or a finally block) once the bitmap has been decoded to avoid leaking file descriptors.

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
keep the change set scoped to a single Android example under examples/

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Merge branch 'main' into main

WangHHY19931001 committed:
Merge branch 'ultralytics:main' into main
